src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -276,112 +281,174 @@ private void CreateEmbeddedAttributesIfNeeded(DiagnosticBag diagnostics)
                     ref _lazyIsReadOnlyAttribute,
                     diagnostics,
-                    AttributeDescription.IsReadOnlyAttribute);
+                    AttributeDescription.IsReadOnlyAttribute,
+                    CreateParameterlessEmbeddedAttributeSymbol);
             }
 
             if ((needsAttributes & EmbeddableAttributes.IsByRefLikeAttribute) != 0)
             {
-                CreateEmbeddedAttributeIfNeeded(
+                CreateAttributeIfNeeded(
                     ref _lazyIsByRefLikeAttribute,
                     diagnostics,
-                    AttributeDescription.IsByRefLikeAttribute);
+                    AttributeDescription.IsByRefLikeAttribute,
+                    CreateParameterlessEmbeddedAttributeSymbol);
             }
 
             if ((needsAttributes & EmbeddableAttributes.IsUnmanagedAttribute) != 0)
             {
-                CreateEmbeddedAttributeIfNeeded(
+                CreateAttributeIfNeeded(
                     ref _lazyIsUnmanagedAttribute,
                     diagnostics,
-                    AttributeDescription.IsUnmanagedAttribute);
+                    AttributeDescription.IsUnmanagedAttribute,
+                    CreateParameterlessEmbeddedAttributeSymbol);
             }
 
             if ((needsAttributes & EmbeddableAttributes.NullableAttribute) != 0)
             {
-                CreateEmbeddedNullableAttributeIfNeeded(
+                CreateAttributeIfNeeded(
                     ref _lazyNullableAttribute,
-                    diagnostics);
+                    diagnostics,
+                    AttributeDescription.NullableAttribute,
+                    CreateNullableAttributeSymbol);
             }
 
             if ((needsAttributes & EmbeddableAttributes.NullableContextAttribute) != 0)
             {
-                CreateEmbeddedNullableContextAttributeIfNeeded(
+                CreateAttributeIfNeeded(
                     ref _lazyNullableContextAttribute,
-                    diagnostics);
+                    diagnostics,
+                    AttributeDescription.NullableContextAttribute,
+                    CreateNullableContextAttributeSymbol);
             }
 
             if ((needsAttributes & EmbeddableAttributes.NullablePublicOnlyAttribute) != 0)
             {
-                CreateEmbeddedNullablePublicOnlyAttributeIfNeeded(
+                CreateAttributeIfNeeded(
                     ref _lazyNullablePublicOnlyAttribute,
-                    diagnostics);
+                    diagnostics,
+                    AttributeDescription.NullablePublicOnlyAttribute,
+                    CreateNullablePublicOnlyAttributeSymbol);
             }
         }
 
-        private void CreateEmbeddedAttributeIfNeeded(
-            ref SynthesizedEmbeddedAttributeSymbol symbol,
+        private SynthesizedEmbeddedAttributeSymbol CreateParameterlessEmbeddedAttributeSymbol(string name, NamespaceSymbol containingNamespace, DiagnosticBag diagnostics)
```

Comment:
>private SynthesizedEmbeddedAttributeSymbol [](start = 8, length = 42)

Consider declaring as:
```
private readonly Func<string, NamespaceSymbol, DiagnosticBag, SynthesizedEmbeddedAttributeSymbol> CreateParameterlessEmbeddedAttributeSymbol(...);
```

Reply:
This doesn't work since the lambda bodies reference instance members, so I'd need to move them to constructor. Not sure if it's worth the effort for saving a few delegate allocations per compilation emit.

I could store the first one in a local since it is used 4 times:



---
In reply to: [302215939](https://github.com/dotnet/roslyn/pull/36854#discussion_r302215939) [](ancestors = 302215939)

Reply:
Makes sense to store the first in a local and leave the methods as methods.

---
In reply to: [302235341](https://github.com/dotnet/roslyn/pull/36854#discussion_r302235341) [](ancestors = 302235341,302215939)